File: Packages/RallyUI/Sources/RallyUI/Components/EmptyState.swift
import SwiftUI
import RallyCore

/// Empty state placeholder view with an illustration, message, and optional
/// call-to-action button.
///
/// Usage:
/// ```swift
/// EmptyState(
///     icon: "calendar.badge.exclamationmark",
///     title: "No Upcoming Events",
///     message: "Check back soon for the next gameday.",
///     actionTitle: "Browse Schools"
/// ) {
///     navigateToSchools()
/// }
/// ```
public struct EmptyState: View {

    // MARK: - Properties

    private let icon: String
    private let title: String
    private let message: String
    private let actionTitle: String?
    private let action: (() -> Void)?

    // MARK: - Init

    /// Creates an empty state view.
    /// - Parameters:
    ///   - icon: SF Symbol name for the illustration.
    ///   - title: Primary heading text.
    ///   - message: Descriptive body text explaining why this state is shown.
    ///   - actionTitle: Optional button label. If `nil`, no button is shown.
    ///   - action: Closure invoked when the optional CTA button is tapped.
    public init(
        icon: String,
        title: String,
        message: String,
        actionTitle: String? = nil,
        action: (() -> Void)? = nil
    ) {
        self.icon = icon
        self.title = title
        self.message = message
        self.actionTitle = actionTitle
        self.action = action
    }

    // MARK: - Body

    public var body: some View {
        VStack(spacing: RallySpacing.md) {
            Spacer()

            // Illustration icon
            Image(systemName: icon)
                .font(.system(size: 56, weight: .light))
                .foregroundStyle(
                    LinearGradient(
                        colors: [RallyColors.gray.opacity(0.6), RallyColors.gray.opacity(0.3)],
                        startPoint: .top,
                        endPoint: .bottom
                    )
                )
                .padding(.bottom, RallySpacing.sm)
                .accessibilityHidden(true)

            // Title
            Text(title)
                .font(RallyTypography.sectionHeader)
                .foregroundStyle(.white)
                .multilineTextAlignment(.center)

            // Message
            Text(message)
                .font(RallyTypography.body)
                .foregroundStyle(RallyColors.gray)
                .multilineTextAlignment(.center)
                .lineSpacing(4)
                .fixedSize(horizontal: false, vertical: true)

            // Optional CTA
            if let actionTitle, let action {
                RallyButton(
                    actionTitle,
                    style: .secondary,
                    size: .medium,
                    isFullWidth: false
                ) {
                    action()
                }
                .padding(.top, RallySpacing.sm)
            }

            Spacer()
        }
        .padding(.horizontal, RallySpacing.xl)
        .frame(maxWidth: .infinity)
        .accessibilityElement(children: .contain)
    }
}

// MARK: - Convenience Initializers

public extension EmptyState {
    /// Empty state for no events.
    static func noEvents(action: (() -> Void)? = nil) -> EmptyState {
        EmptyState(
            icon: "calendar.badge.exclamationmark",
            title: "No Upcoming Events",
            message: "There are no events scheduled right now. Check back soon for the next gameday!",
            actionTitle: action != nil ? "Refresh" : nil,
            action: action
        )
    }

    /// Empty state for no rewards.
    static func noRewards(action: (() -> Void)? = nil) -> EmptyState {
        EmptyState(
            icon: "gift",
            title: "No Rewards Available",
            message: "New rewards are added regularly. Keep earning points and check back soon!",
            actionTitle: action != nil ? "Browse Events" : nil,
            action: action
        )
    }

    /// Empty state for an empty leaderboard.
    static func noLeaderboard() -> EmptyState {
        EmptyState(
            icon: "trophy",
            title: "Leaderboard Coming Soon",
            message: "Complete activations during the event to see your ranking here."
        )
    }

    /// Empty state for network errors.
    static func networkError(action: (() -> Void)? = nil) -> EmptyState {
        EmptyState(
            icon: "wifi.slash",
            title: "Connection Issue",
            message: "We could not load this content. Please check your connection and try again.",
            actionTitle: "Try Again",
            action: action
        )
    }
}

// MARK: - Preview

#Preview("Empty States") {
    ScrollView {
        VStack(spacing: 48) {
            EmptyState(
                icon: "calendar.badge.exclamationmark",
                title: "No Upcoming Events",
                message: "There are no events scheduled right now. Check back soon!",
                actionTitle: "Browse Schools"
            ) {}

            Divider().background(RallyColors.gray.opacity(0.3))

            EmptyState.networkError {}

            Divider().background(RallyColors.gray.opacity(0.3))

            EmptyState.noLeaderboard()
        }
    }
    .background(RallyColors.navy)
    .environment(ThemeEngine())
}

File: Android/ui/src/main/java/com/rally/app/ui/components/EmptyState.kt
package com.rally.app.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.CardGiftcard
import androidx.compose.material.icons.filled.EmojiEvents
import androidx.compose.material.icons.filled.WifiOff
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.rally.app.ui.theme.RallyColors
import com.rally.app.ui.theme.RallySpacing
import com.rally.app.ui.theme.RallyTypography

// ---------------------------------------------------------------------------
// EmptyState
// Icon + title + message + optional CTA button.
// Provides convenience factory functions for common empty states:
// noEvents, noRewards, noLeaderboard, networkError.
// ---------------------------------------------------------------------------

/**
 * Empty state placeholder view.
 *
 * @param icon         [ImageVector] illustration displayed at the top.
 * @param title        Primary heading text.
 * @param message      Descriptive body text.
 * @param modifier     Optional [Modifier].
 * @param actionTitle  Optional button label. If `null`, no button is rendered.
 * @param onAction     Callback invoked when the optional CTA is tapped.
 */
@Composable
fun EmptyState(
    icon: ImageVector,
    title: String,
    message: String,
    modifier: Modifier = Modifier,
    actionTitle: String? = null,
    onAction: (() -> Unit)? = null,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = RallySpacing.xl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Spacer(modifier = Modifier.weight(1f))

        // Illustration icon
        Icon(
            imageVector = icon,
            contentDescription = null, // decorative
            tint = RallyColors.Gray.copy(alpha = 0.5f),
            modifier = Modifier
                .size(56.dp)
                .padding(bottom = RallySpacing.sm),
        )

        // Title
        Text(
            text = title,
            style = RallyTypography.sectionHeader,
            color = androidx.compose.ui.graphics.Color.White,
            textAlign = TextAlign.Center,
        )

        Spacer(modifier = Modifier.height(RallySpacing.sm))

        // Message
        Text(
            text = message,
            style = RallyTypography.body,
            color = RallyColors.Gray,
            textAlign = TextAlign.Center,
            lineHeight = RallyTypography.body.lineHeight,
        )

        // Optional CTA
        if (actionTitle != null && onAction != null) {
            Spacer(modifier = Modifier.height(RallySpacing.md))
            RallyButton(
                title = actionTitle,
                onClick = onAction,
                style = RallyButtonStyle.Secondary,
                size = RallyButtonSize.Medium,
                isFullWidth = false,
            )
        }

        Spacer(modifier = Modifier.weight(1f))
    }
}

// MARK: - Convenience Factories

object EmptyStates {
    /** Empty state for no events. */
    @Composable
    fun NoEvents(modifier: Modifier = Modifier, onRefresh: (() -> Unit)? = null) {
        EmptyState(
            icon = Icons.Filled.CalendarMonth,
            title = "No Upcoming Events",
            message = "There are no events scheduled right now. Check back soon for the next gameday!",
            modifier = modifier,
            actionTitle = if (onRefresh != null) "Refresh" else null,
            onAction = onRefresh,
        )
    }

    /** Empty state for no rewards. */
    @Composable
    fun NoRewards(modifier: Modifier = Modifier, onBrowse: (() -> Unit)? = null) {
        EmptyState(
            icon = Icons.Filled.CardGiftcard,
            title = "No Rewards Available",
            message = "New rewards are added regularly. Keep earning points and check back soon!",
            modifier = modifier,
            actionTitle = if (onBrowse != null) "Browse Events" else null,
            onAction = onBrowse,
        )
    }

    /** Empty state for an empty leaderboard. */
    @Composable
    fun NoLeaderboard(modifier: Modifier = Modifier) {
        EmptyState(
            icon = Icons.Filled.EmojiEvents,
            title = "Leaderboard Coming Soon",
            message = "Complete activations during the event to see your ranking here.",
            modifier = modifier,
        )
    }

    /** Empty state for network errors. */
    @Composable
    fun NetworkError(modifier: Modifier = Modifier, onRetry: (() -> Unit)? = null) {
        EmptyState(
            icon = Icons.Filled.WifiOff,
            title = "Connection Issue",
            message = "We could not load this content. Please check your connection and try again.",
            modifier = modifier,
            actionTitle = "Try Again",
            onAction = onRetry,
        )
    }
}

// MARK: - Previews

@Preview(name = "EmptyState - Variants", showBackground = true, backgroundColor = 0xFF131B2E)
@Composable
private fun EmptyStatePreview() {
    Surface(color = RallyColors.Navy) {
        Column(
            modifier = Modifier
                .padding(RallySpacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.xxxl),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            EmptyState(
                icon = Icons.Filled.CalendarMonth,
                title = "No Upcoming Events",
                message = "There are no events scheduled right now. Check back soon!",
                actionTitle = "Browse Schools",
                onAction = {},
            )

            Divider(color = RallyColors.Gray.copy(alpha = 0.3f))

            EmptyStates.NetworkError(onRetry = {})

            Divider(color = RallyColors.Gray.copy(alpha = 0.3f))

            EmptyStates.NoLeaderboard()
        }
    }
}
